### PR TITLE
handeled the mongoose connection error on backend and made necessay c…

### LINKED ISSUE #24: Handle Mongoose Connection Failures Gracefully
#### **Description** 
Mongoose fails to connect to the database, the application does not currently handle the error, causing the app to crash unexpectedly. One common reason is that you're trying to access the database from an IP that isn't whitelisted. This leads to poor user experience on the frontend and unclear debugging information on the backend. To address this, we need to implement both frontend and backend improvements to ensure meaningful feedback, error handling and stability during database outages.

---

#### **Steps to Reproduce**  
1. Stop the MongoDB server/use an invalid database URI/remove your current ip address from network access list on security tab of mongodb console.
2. Start your local server.
3. Try to access the `/api/Tickets` endpoint via the frontend.
4. Observe that the error is not handled effectively on either of the backend and frontend and the app breaks.

---

### **1. Backend Enhancements**  

- **Handle Database Connection Errors Gracefully**:  
  - Wrap the Mongoose connection logic in a `try-catch` block and handle connection errors.  
  - Provide meaningful HTTP responses to the client when a database connection fails (e.g., return `503 Service Unavailable`).

---

### **2. Frontend Enhancements**

- **Graceful Fallback UI**:  
  Show a user-friendly message when the backend is unable to fetch data due to database connection issues.  

    ```javascript
    // app/page.jsx

    const data = await getTickets();

    if (!data || !data?.tickets) {
      return (
        <div className="p-5">
          <h2 className="text-red-500">
            Unable to load tickets. Please try again later.
          </h2>
        </div>
      );
    }
    
    const tickets = data.tickets;
    ```

### PR DESCRIPTION
**Pull request by anshkarwasra**
**closes #24 **

**type of change**

- [X] bug fix

**description of change** 
handled the undefined mongoruri on reload and added no ticket found page on frontend in case there are no tickets 
still a long warning was showing on app reload (tried to fix that but wasn't able to :/ )

**implementation details**

added a try catch statement in `Tickets.js`  and returned null if any error happens while connecting to the mongoUri in `route.js`
made ticket not found response if tickets are null

### CODE REVIEW

File: package.json [hunk 1]
```diff
@@ -18,6 +18,7 @@
     "mongoose": "^7.4.3",
     "next": "13.4.13",
     "postcss": "8.4.27",
+    "punycode": "^2.3.1",
```
Comment (AyushDharDubey) [inline package.json:21]: always ensure that your commits stay focused on the objective of the PR. Refrain from including unrelated changes in order to keep the PR clear and on track.

File: app/models/Ticket.js [hunk 2]
```diff
@@ -1,23 +1,40 @@
...
-    category: String,
-    priority: Number,
-    progress: Number,
-    status: String,
-    active: Boolean,
-  },
-  {
-    timestamps: true,
+import mongoose, { mongo, Schema } from "mongoose";
+let Ticket;
```
Comment (AyushDharDubey) [inline app/models/Ticket.js:2]: Can you explain what benefit would this change provide over using `const Ticket`.
I am under the impression that we never want to update the Ticket object, do you believe otherwise?

File: app/page.jsx [hunk 3]
```diff
@@ -1,35 +1,61 @@
...
+const uniqueCategories = [
     ...new Set(tickets?.map(({ category }) => category)),
   ];
 
   return (
-    <div className="p-5">
+    <div>
+      {loading ? (
+        <p>Loading...</p>
+      ) : tickets.length === 0 ? (
```
Comment (AyushDharDubey) [inline app/page.jsx:51]: This is issue #14 which needs to be addressed by creating a separate PR

File: app/page.jsx [hunk 4]
```diff
@@ -1,35 +1,61 @@
-import React from "react";
-import TicketCard from "./(components)/TicketCard";
+"use client"
+import { useState, useEffect } from "react";
+import TicketCard from "./(components)/TicketCard"; // Adjust the import according to your file structure
 
-const getTickets = async () => {
```
Comment (AyushDharDubey) [inline app/page.jsx:4]: why do you believe `getTickets` is required to be declared inside `useEffect`?

File: app/models/Ticket.js [hunk 5]
```diff
@@ -1,23 +1,40 @@
...
+try {
+  if (process.env.MONGODB_URI === undefined) {
```
Comment (AyushDharDubey) [inline app/models/Ticket.js:4]: good! this can handle the case when the `MONGODB_URI` is not set at all

**Comment by AyushDharDubey** [comment]
> still a long warning was showing on app reload (tried to fix that but wasn't able to :/ )

that's why contributor's are instructed to discuss their approach and seek guidance on issue page first

Ponder over above reviews and look into following untouched issues
1. you failed to address the edge case when the env var MONGODB_URI is set but has illegal value
2. you failed to address the case when even-though the variable would be correctly defined, but your current ip address might not be added in the **network access** list of mongodb clusters which would restrict your network from establishing a db connection.

File: app/page.jsx (hunk 4 above)
Comment (anshkarwasra) [reply app/page.jsx:4]: If there is any update in ticket database that should reflect in the application in sync with it so I believe that featching the tickets every time the page renders keep the frontend updated

**Comment by anshkarwasra** [comment]
First of all i didn't understand why the mongouri env variable become undefined when I reload the website after starting the app

File: app/page.jsx (hunk 3 above)
Comment (anshkarwasra) [reply app/page.jsx:51]: Apologies for that I thought it would be nice and clean to add a loading State

File: app/models/Ticket.js (hunk 2 above)
Comment (anshkarwasra) [reply app/models/Ticket.js:2]: I just did it so that in Rout.js I can send a specific response from server when server was unable to  featch tickets ( by raising an exception that tickets are null further handling it and sending a ticket not found response to frontend)

File: app/api/Tickets/[id]/route.js [hunk 6]
```diff
@@ -4,8 +4,16 @@ import { NextResponse } from "next/server";
 export async function GET(request, { params }) {
   const { id } = params;
 
+  try{
+      if(Ticket === null) {return NextResponse.json({ message: "Ticket not found" }, { status: 404 });
```
[the remark this answers is not in the document]
Comment (anshkarwasra) [reply app/api/Tickets/[id]/route.js:8]: Sorry to say but if we weren't able to connect to DB then how tickets can be featched (that's why set them to null if fails to connect to DB and further handling that case)

File: package.json (hunk 1 above)
Comment (AyushDharDubey) [reply package.json:21]: revert this change

File: app/page.jsx (hunk 3 above)
Comment (AyushDharDubey) [reply app/page.jsx:51]: you need to revert this

File: app/page.jsx (hunk 4 above)
Comment (AyushDharDubey) [reply app/page.jsx:4]: that could have been achieved without introducing this change, the getTicket is already called each time the we hit the dashboard
it would be nice if you can revert this one

File: app/api/Tickets/[id]/route.js (hunk 6 above)
[the remark this answers is not in the document]
Comment (AyushDharDubey) [reply app/api/Tickets/[id]/route.js:8]: > if we weren't able to connect to DB then how tickets can be featched

that is also an edge case which needs to be handled gracefully,
in the said case, we need to give separate feedback to the user that the application couldn't establish successful connections with the cluster.

[the remark this answers is not in the document]
Comment (anshkarwasra) [reply app/api/Tickets/[id]/route.js:8]: Okay I will provide a different response for that case and adjust the ui accordingly

File: app/page.jsx (hunk 4 above)
Comment (anshkarwasra) [reply app/page.jsx:4]: But wouldn't use effect would run it as a side effect improve our site's proformance (just being curious)

**Comment by AyushDharDubey** [comment]
Hi @anshkarwasra,
are you still working on the issue?

If you need more time, please feel free to let us know. If you require assistance, kindly inform us before 28th Dec, 2359 hours, so we can explore alternative solutions if necessary.

if we don't hear back by then, unfortunately, we may have to close the PR to give others a chance to contribute.

hope you will understand, looking forward to hearing from you.

**Comment by anshkarwasra** [comment]
Hey sorry for the inactivity I was busy on some other projects would try to push the by today
